### PR TITLE
Support custom s3 api endpoints

### PR DESCRIPTION
**Pull request by easel**
Rebased #619 against latest master.

### CODE REVIEW

**Comment by easel** [comment]
I've added some documentation and adjusted the s3 integration tests to work against Minio. Aside from this patch, getting minio to pass the tests requires falling back to url encoding the responses from the upload calls. 

The regular test suite still passes for me, and unless I'm missing something I don't believe any of the changes should change the code path unless an endpoint url is specified. Happy to address any other feedback.

@takashima0411 @ktoso any thoughts?

**Comment by ktoso** [comment]
Sorry for the delay, family time back home until New Years :)
Will review then!

**Comment by easel** [comment]
Thanks @ktoso, happy New Year!

**Comment by ennru** [comment]
Does this replace #619 ? Please close it if so.

**Comment by easel** [comment]
Yes, it's includes all of #619 but includes fixes, additional capabilities and tests. I didn't create #619 so I don't believe I can update it or delete it, hence the new PR.

**Comment by francisdb** [comment]
Please keep the `AwsS3IntegrationSpec` as the same previous refactor to `S3IntegrationSpec` was no longer doing proper validation of all AWS S3 cases.

I personally think it's better to have a specific test for each service as they each have their edge cases and url path issues.

**Comment by easel** [comment]
The AWSS3IntegrationSpec is still there, I just lifted the test methods up to a trait so I could inject the differing config for this patch. The tests are all identical to before and they all pass against minio -- as far as I am concerned if a provider claims to be s3 compatible it should pass the same integration suite.

**Comment by francisdb** [comment]
Seems I was a bit too fast, it's the `S3NoMock` test that is totally broken and that was why I re-introduced an old version as the AWSS3IntegrationSpec.

File: s3/src/main/resources/reference.conf [hunk 1]
```diff
@@ -73,4 +73,7 @@ akka.stream.alpakka.s3 {
   # When using virtual hosted–style buckets with SSL, the S3 wild card certificate only matches buckets that do not contain periods.
   # Buckets containing periods will lead to certificate errors. In those cases it's useful to enable path-style access.
   path-style-access = false
+
+  # Custom endpoint url, used for alternate s3 implementations
+  endpoint-url = null
```
Comment (francisdb) [inline s3/src/main/resources/reference.conf]: I wonder if this is discouraged

> If you do need a setting to be optional: checking hasPath() in advance should be the same amount of code (in Java) as checking for null afterward, without the risk of NullPointerException when you forget. In Scala, you could write an enrichment class like this to use the idiomatic Option syntax:
https://github.com/lightbend/config#how-to-handle-defaults

File: s3/src/test/java/akka/stream/alpakka/s3/javadsl/DocSnippets.java [hunk 2]
```diff
@@ -46,7 +47,8 @@ public String getRegion() {
                 Some.apply(proxy),
                 credentials,
                 regionProvider,
-                true
+                true,
+                Option.<String>none().asScala() //FIXME
```
Comment (francisdb) [inline s3/src/test/java/akka/stream/alpakka/s3/javadsl/DocSnippets.java]: Still needs fix to allow a java Optional?

**Comment by easel** [comment]
Yeah, I couldn't figure that one out at all =p The AWSS3IntegrationSpec worked ok, I just added some delete calls to make it idempotent.

File: s3/src/test/java/akka/stream/alpakka/s3/javadsl/S3ClientTest.java [hunk 3]
```diff
@@ -58,7 +59,8 @@ public String getRegion() throws SdkClientException {
             Some.apply(proxy),
             credentials,
             regionProvider("us-east-1"),
-            false
+            false,
+            akka.japi.Option.<String>none().asScala() //FIXME
```
Comment (francisdb) [inline s3/src/test/java/akka/stream/alpakka/s3/javadsl/S3ClientTest.java]: Still needs fix to allow a java Optional?

File: s3/src/main/resources/reference.conf (hunk 1 above)
Comment (easel) [reply s3/src/main/resources/reference.conf]: Perhaps? I guess the the alternate solution would be to include the key, commented out. It seems like it should be documented in the reference.conf file in some capacity or another.

File: s3/src/test/java/akka/stream/alpakka/s3/javadsl/DocSnippets.java (hunk 2 above)
Comment (easel) [reply s3/src/test/java/akka/stream/alpakka/s3/javadsl/DocSnippets.java]: I'm not sure I'm following. Are you suggesting we could simplify that last line?

Comment (francisdb) [reply s3/src/test/java/akka/stream/alpakka/s3/javadsl/DocSnippets.java]: I saw the FIXME so my guess was you wanted something cleaner?

Comment (easel) [reply s3/src/test/java/akka/stream/alpakka/s3/javadsl/DocSnippets.java]: Ah, no, sorry, I carried that over from master when I merged this branch with master. I'm not sure what the intent was... off-hand I'd guess they'd prefer to provide something smarter than a hardcoded none but weren't sure what.

**Comment by ennru** [comment]
@easel and @francisdb can you get this one ready for merge and release? I'd like to get 0.16 out on Thursday.

**Comment by francisdb** [comment]
> francisdb approved these changes 5 days ago 
just added some small remarks

File: s3/src/test/java/akka/stream/alpakka/s3/javadsl/S3ClientTest.java (hunk 3 above)
Comment (ennru) [reply s3/src/test/java/akka/stream/alpakka/s3/javadsl/S3ClientTest.java]: What about `scala.Option.empty()`?
We should add Java API to S3Settings to bridge the case class. (see e.g. JmsSourceSettings), but it doesn't need to be done in the PR.

**Comment by takashima0411** [comment]
@easel sorry for delay, and thank you for the patches. 
Minio seems good  :+1:

**Comment by easel** [comment]
Thanks all for the feedback. I've rebased against master and adjusted things per earlier comments.

- Removed the remaining references to baseUrl
- Default to a commented out endpoint-url key, and added tests that ensure missing keys and resetting to null both are handled as None
- Switched java options to scala.Option.empty(), which compiles. I don't use java api's anymore so I can't say if it's a better solution personally but the code is cleaner!

**Comment by ennru** [comment]
Fixes #181 

Thank you for you contribution! Keep going!

**Comment by easel** [comment]
Thanks @ennru !

**Comment by ennru** [comment]
Cheers. We 'll release 0.16 on Thursday.

**Comment by l15k4** [comment]
Thanks guys, it seems to be working with https://github.com/findify/s3mock but I still cannot figure out why multipart upload doesn't work on https://github.com/jubos/fake-s3 ...

I set it this way : 
```
      endpointUrl = Some(s"http://localhost:$randomPort"),
      pathStyleAccess = true
```

```
akka.http.scaladsl.unmarshalling.Unmarshaller$UnsupportedContentTypeException: Unsupported Content-Type, supported: application/xml, application/octet-stream
	at akka.http.scaladsl.unmarshalling.Unmarshaller$UnsupportedContentTypeException$.apply(Unmarshaller.scala:158)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$EnhancedFromEntityUnmarshaller$.$anonfun$forContentTypes$3(Unmarshaller.scala:114)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$$anon$1.apply(Unmarshaller.scala:58)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$EnhancedUnmarshaller$.$anonfun$mapWithInput$3(Unmarshaller.scala:91)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$$anon$1.apply(Unmarshaller.scala:58)
	at akka.http.scaladsl.unmarshalling.Unmarshaller.$anonfun$transform$3(Unmarshaller.scala:23)
	at akka.http.scaladsl.unmarshalling.Unmarshaller$$anon$1.apply(Unmarshaller.scala:58)
	at akka.http.scaladsl.unmarshalling.Unmarshal.to(Unmarshal.scala:25)
	at akka.stream.alpakka.s3.impl.S3Stream.$anonfun$initiateMultipartUpload$3(S3Stream.scala:216)
```

**Comment by bandrzejczak** [comment]
@l15k4 Yes it does. While using standard S3 client I had to disable chunked encoding to make it work with fake-s3. It still allows for sending your document in multiple chunks though, right?
